arch/cpu/gecko/dev/gpio-hal-arch.c

```diff
@@ -0,0 +1,177 @@
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * \addtogroup gecko
+ * @{
+ *
+ * \addtogroup gecko-dev Device drivers
+ * @{
+ *
+ * \addtogroup gecko-gpio GPIO HAL driver
+ * @{
+ *
+ * \file
+ *     GPIO HAL implementation for the Gecko
+ * \author
+ *     Yago Fontoura do Rosario <yago.rosario@hotmail.com.br>
+ *
+ */
+/*---------------------------------------------------------------------------*/
+#include "contiki.h"
+
+#include "dev/gpio-hal.h"
+
+#include "em_gpio.h"
+#include "em_core.h"
+#include "em_cmu.h"
+#include "gpiointerrupt.h"
+/*---------------------------------------------------------------------------*/
+typedef struct {
+  gpio_hal_port_t port;
+  gpio_hal_pin_t pin;
+  gpio_hal_pin_cfg_t cfg;
+} gpio_hal_cache_t;
+/*---------------------------------------------------------------------------*/
+#define GPIO_HAL_CACHE_SIZE 32
+/*---------------------------------------------------------------------------*/
+static gpio_hal_cache_t gpio_hal_cache[GPIO_HAL_CACHE_SIZE] = { 0 };
+/*---------------------------------------------------------------------------*/
+static void
+gpio_hal_arch_callback(uint8_t interrupt_no, void *ctx)
+{
+  (void)ctx;
+  gpio_hal_event_handler(gpio_hal_cache[interrupt_no].port,
+                         1 << gpio_hal_cache[interrupt_no].pin);
+}
+/*---------------------------------------------------------------------------*/
+static uint8_t
+gpio_hal_arch_cache_lookup(gpio_hal_port_t port, gpio_hal_pin_t pin)
+{
+  /* Logic extracted from GPIOINT_CallbackRegisterExt */
+#if defined(_GPIO_EXTIPINSELL_MASK)
+  uint32_t intToCheck;
```

Comment:
A general remark is that code checking tools such as Clang static analyzer produces a lot of noise from variables that are not in the innermost scope possible. Everything should be compiled as C99 now, so the compiler should allow you to declare variables in the innermost scope possible.

For this particular code, I think `intToCheck` can be both declared and defined inside the body of the if, which will also save a few cycles in the loop.

Reply:
Yeah, if it's imported sources it's better to leave as is so it's easy to diff against "upstream" in the future.